manticore/core/memory.py

```diff
@@ -1071,6 +1137,210 @@ def _try_get_solutions(self, address, size, access, max_solutions=0x1000, force=
+        fdata = fdata[offset:]
+        fdata = fdata.ljust(size, b'\0')
+
+        for i in range(size):
+            Memory.write(self, addr + i, chr(fdata[i]), force=True)
+
+        logger.debug('New file-memory map @%x size:%x', addr, size)
+
+        return addr
+
+    def _deref_can_succeed(self, map, address, size):
+        if not issymbolic(address):
+            return address >= map.start and address + size < map.end
+        else:
+            # FIXME is address + size < map.end a bug? address + size is first oob of the access. address + size <= map.end?
+            constraint = Operators.AND(address >= map.start, address + size < map.end)
+            return solver.can_be_true(self.constraints, constraint)
+
+    def _import_concrete_memory(self, from_addr, to_addr):
+        """
+        for each address in this range need to read from concrete and write to symbolic
+        it's possible that there will be invalid/unmapped addresses in this range. need to skip to next map if so
+        also need to mark all of these addresses as now in the symbolic store
+
+        :param int from_addr:
+        :param int to_addr:
+        :return:
+        """
+        logger.debug("Importing concrete memory: {:x} - {:x} ({} bytes)".format(from_addr, to_addr, to_addr - from_addr))
+
+        for m in self.maps:
+            span = interval_intersection(m.start, m.end, from_addr, to_addr)
+            if span is None:
+                continue
+
+            start, stop = span
+
+            for addr in range(start, stop):
+                if addr in self.backed_by_symbolic_store:
+                    continue
+
+                self.backing_array[addr] = Memory.read(self, addr, 1)[0]
+                self.backed_by_symbolic_store.add(addr)
+
+    def _map_deref_expr(self, map, address, size):
+        return Operators.AND(
+            # address >= map.start,
+            # address + size < map.end)
+            Operators.UGE(address, map.start),
+            Operators.ULT(address, map.end))
+
+    def _reachable_range(self, sym_address, size):
+        addr_min, addr_max = solver.minmax(self.constraints, sym_address)
+        return addr_min, addr_max + size - 1
+
+    def valid_ptr(self, address):
+        assert issymbolic(address)
+
+        expressions = [self._map_deref_expr(m, address, 1) for m in self._maps]
+        valid = functools.reduce(Operators.OR, expressions)
+
+        return valid
+
+    def invalid_ptr(self, address):
+        return Operators.NOT(self.valid_ptr(address))
+
+    def read(self, address, size, force=False):
+
+        # if address is symbolic, self.valid_ptr(address) ? ?
+        access_min, access_max = self._reachable_range(address, size)
+
+        if issymbolic(address):
+            self._import_concrete_memory(access_min, access_max)
+
+        retvals = []
+        addrs_to_access = [address + i for i in range(size)]
+        for addr in addrs_to_access:
+            from_array = False
```

Comment:
This means whether memory (address) is backed by a an array of symbolic values? Or something different?

The `if ... elif` is a bit unclear and it might be good to add some explanation here as a comment, e.g. why do we care if given memory has write permission?